File: android-wife-assistant/app/src/test/java/com/wifeassistant/OccasionsTest.kt
package com.wifeassistant

import com.wifeassistant.data.OccasionConfig
import com.wifeassistant.data.Occasions
import com.wifeassistant.data.PersonOccasion
import com.wifeassistant.data.Recipient
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

// اختبار مطابقة المناسبات (Occasions.match) — نظير occasions.test.js.
class OccasionsTest {

    @Test
    fun `مناسبة ثابتة MM-DD بتتطابق`() {
        val res = Occasions.match(
            listOf(OccasionConfig("fixed", date = "08-24", label = "عيد ميلاد مراتي")),
            ymd = "2026-08-24", mmdd = "08-24",
        )
        assertEquals("عيد ميلاد مراتي", res?.label)
    }

    @Test
    fun `مناسبة يدوية dates بتتطابق`() {
        val res = Occasions.match(
            listOf(OccasionConfig("manual", dates = listOf("2000-01-01", "2026-08-25"), label = "المولد")),
            ymd = "2026-08-25", mmdd = "08-25",
        )
        assertEquals("المولد", res?.label)
    }

    @Test
    fun `الـ placeholders بتتجاهل`() {
        val res = Occasions.match(
            listOf(
                OccasionConfig("fixed", date = "MM-DD", label = "x"),
                OccasionConfig("manual", dates = listOf("YYYY-MM-DD"), label = "y"),
            ),
            ymd = "2026-08-25", mmdd = "08-25",
        )
        assertNull(res)
    }

    @Test
    fun `المناسبة المعطّلة بتتجاهل`() {
        val res = Occasions.match(
            listOf(OccasionConfig("fixed", date = "08-24", label = "x", enabled = false)),
            ymd = "2026-08-24", mmdd = "08-24",
        )
        assertNull(res)
    }

    @Test
    fun `يوم من غير مناسبة بيرجع null`() {
        val res = Occasions.match(
            listOf(OccasionConfig("fixed", date = "01-01", label = "x")),
            ymd = "2026-07-05", mmdd = "07-05",
        )
        assertNull(res)
    }

    @Test
    fun `مناسبة الشخص الخاصة بتتطابق باليوم`() {
        val r = Recipient(
            id = "1", name = "سارة", relation = "daughter",
            occasions = listOf(PersonOccasion("عيد ميلاد", "08-24")),
        )
        assertEquals("عيد ميلاد", Occasions.recipientOccasionToday(r, "08-24")?.label)
        assertNull(Occasions.recipientOccasionToday(r, "01-01"))
    }
}

File: telegram-wife-assistant/test/occasions.test.js
const { test, afterEach } = require('node:test');
const assert = require('node:assert');

const config = require('../config');
const util = require('../util');
const { getTodaysOccasion } = require('../occasions');

const { ymd, mmdd } = util.todayParts();

// نرجّع نسخة نضيفة من occasions بعد كل اختبار عشان الاختبارات ما تأثّرش على بعض.
const original = JSON.parse(JSON.stringify(config.occasions));
afterEach(() => {
  config.occasions = JSON.parse(JSON.stringify(original));
});

test('مناسبة ثابتة (fixed MM-DD) بتتطابق النهاردة', () => {
  config.occasions = { anniv: { type: 'fixed', date: mmdd, label: 'عيد جوازنا' } };
  assert.strictEqual(getTodaysOccasion()?.label, 'عيد جوازنا');
});

test('مناسبة يدوية (manual dates[]) بتتطابق النهاردة', () => {
  config.occasions = {
    eid: { type: 'manual', dates: ['2000-01-01', ymd], label: 'عيد' },
  };
  assert.strictEqual(getTodaysOccasion()?.label, 'عيد');
});

test('الـ placeholders (MM-DD / YYYY-MM-DD) بتتجاهل', () => {
  config.occasions = {
    a: { type: 'fixed', date: 'MM-DD', label: 'x' },
    b: { type: 'manual', dates: ['YYYY-MM-DD'], label: 'y' },
  };
  assert.strictEqual(getTodaysOccasion(), null);
});

test('يوم من غير مناسبة بيرجّع null', () => {
  // تاريخ مستحيل يطابق النهاردة
  config.occasions = { a: { type: 'fixed', date: '01-01', label: 'x' } };
  const res = getTodaysOccasion();
  // ممكن يصادف 1 يناير، فنتحقق منطقياً:
  if (mmdd !== '01-01') assert.strictEqual(res, null);
  else assert.strictEqual(res?.label, 'x');
});
